### PR TITLE
feat: Add CSV ingestion support for Frames

### LINKED ISSUE #70: Support for Videos / frames on OM

### PR DESCRIPTION
**Pull request by ad-claw000**
Closes #70

This PR adds a `FrameDataCSV` class which simply inherits from `ImageDataCSV` and overrides the command to `AddFrame`. It also registers `IngestType.FRAME` in the CLI, closing the loop on Frame OM and ingestion support without duplicating the complex loading/validation logic of Images.

### CODE REVIEW

**Review by luisremis — CHANGES_REQUESTED** [review]
add testing to the new FrameDataCSV.
add Frame to the OM, analogous to how Image and Video work.

File: aperturedb/FrameDataCSV.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+from aperturedb.ImageDataCSV import ImageDataCSV
+
+class FrameDataCSV(ImageDataCSV):
+    def __init__(self, *args, **kwargs):
+        self.command = "AddFrame"
+        super().__init__(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (ad-claw000) [reply aperturedb/FrameDataCSV.py]: Overrode `get_indices()` in `FrameDataCSV` to correctly target the `_Frame` class, and adjusted the formatting to satisfy PEP 8/autopep8. See commit fd0c808.

File: test/test_FrameDataCSV.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    with tempfile.NamedTemporaryFile(suffix=".csv", mode="w", delete=False) as f:
+        f.write("url,id\nhttp://example.com/frame.jpg,1\n")
+        f.close()
+
+        try:
+            # We don't actually need the image since check_image=False
+            frame_data = FrameDataCSV(f.name, check_image=False)
+
+            assert frame_data.command == "AddFrame", f"Expected AddFrame, got {
+                frame_data.command}"
```
[the remark this answers is not in the document]
Comment (ad-claw000) [reply test/test_FrameDataCSV.py]: Fixed the syntax error by shortening the assertion line so autopep8 doesn't break the f-string. See commit 803e0fa.

**Comment by ad-claw000** [comment]
Addressed the latest Copilot reviewer comments in commit cce1bc7 (added backward-compatibility shim for `Frames` in `Images.py`, added docstring to `FrameDataCSV`, and removed unused imports in the test module).

File: aperturedb/Images.py [hunk 3]
```diff
@@ -1003,18 +1003,5 @@ def get_properties(self, prop_list: Iterable[str] = []) -> Dict[str, Any]:
...
-
-    Args:
-        client: The database connector, perhaps as returned by `CommonLibrary.create_connector`
-    """
-    db_object = ObjectType.FRAME
-
-    def __init__(self, client, batch_size=100, response=None, **kwargs):
-        super().__init__(client, batch_size=batch_size, response=response, **kwargs)
+# Shim for backward compatibility
+from aperturedb.Frames import Frames  # noqa: F401
```
[the remark this answers is not in the document]
Comment (ad-claw000) [reply aperturedb/Images.py]: Replaced the module-level import with a `__getattr__` shim so that `Frames` is imported lazily on access, avoiding the circular import as requested. See commit e0886c5.

File: aperturedb/DataModels.py [hunk 4]
```diff
@@ -70,9 +70,11 @@ class PolygonDataModel(IdentityDataModel):
     type = ObjectType.POLYGON
 
 
-class FrameDataModel(IdentityDataModel):
+class FrameDataModel(BlobDataModel):
     """Frame data model for ApertureDB.
     """
+    url: Annotated[Optional[str], Field(
+        title="URL", description="URL to file, http, s3 or gs resource")] = None
     type = ObjectType.FRAME
```
[the remark this answers is not in the document]
Comment (ad-claw000) [reply aperturedb/DataModels.py:76]: Removed the overridden `url` field from `FrameDataModel` since it's already inherited from `BlobDataModel`. See commit 17ee930.

**Comment by ad-claw000** [comment]
Addressed feedback from @luisremis:

- Added testing for `FrameDataCSV` in `test_FrameDataCSV.py`.
- Added `Frames` to the Object Model (analogous to `Images` and `Videos`) and added corresponding tests in `test_Frames.py`.

Updated in commit f393910

**Comment by ad-claw000** [comment]
Addressed feedback from @copilot-pull-request-reviewer[bot]:

- Fixed the autopep8 line formatting issues in `test_FrameDataCSV.py` and `Frames.py`.
- Reverted the `url` override in `FrameDataModel` to avoid Pydantic field redefinition errors.

Updated in commits f393910 and 17ee930

**Comment by ad-claw000** [comment]
Addressed the latest Copilot reviewer comment in commit c560a56 by explicitly defining `__all__` in `Images.py` so that `Frames` is exported during `import *`.

File: aperturedb/DataModels.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ad-claw000) [reply aperturedb/DataModels.py:76]: Reverted the accidental commit that changed `FrameDataModel`'s base class back to `IdentityDataModel`. It now correctly inherits from `BlobDataModel` as requested. See commit 8d99ccf.

**Comment by ad-claw000** [comment]
Addressed the latest Copilot reviewer comment in commit 83878fd by caching the lazily-imported `Frames` class in `globals()` to prevent repeated lookups and properly expose it in the module namespace.

**Comment by ad-claw000** [comment]
Addressed the latest review feedback in commit 087f510:
- Updated `FrameDataCSV` to use `ObjectType.FRAME.value`.
- Added backward compatibility import test for `Frames`.
- Documented `FrameDataModel` and added a test for its required fields.
- Cleaned up redundant file closure in `test_FrameDataCSV.py`.

**Comment by ad-claw000** [comment]
Addressed the suppressed Copilot reviewer comments in commit b3d4e59:
- `ImageDataCSV.py`: Added fallback to `'AddImage'` if `getattr(self, 'command')` returns a falsy value like `None`.
- `Images.py`: Explicitly defined `__all__` to prevent exporting unwanted internal/imported symbols and avoid dynamic generation via `globals()`.

**Comment by ad-claw000** [comment]
Addressed the latest Copilot reviewer comments in commit b34b3cc:
- `ImageDataCSV.py`: Improved the fallback logic for `self.command` to ensure explicit but falsy values aren't unintentionally overridden.
- `Images.py`: Added a `TYPE_CHECKING` guarded import for `Frames` and explicitly defined `__dir__` to improve static type checker and IDE support.
- `test_cli_ingest.py`: Added a unit test to verify that `from_csv` properly routes `IngestType.FRAME` to the new `FrameDataCSV` parser.

**Comment by ad-claw000** [comment]
Addressed the suppressed Copilot reviewer comments in commit a50b275:
- `Images.py`: Defined `__all__` before `__dir__` to avoid a potential `NameError` when `dir()` is invoked during module initialization.
- `test_cli_ingest.py`: Removed the unused `pytest` import.

**Comment by ad-claw000** [comment]
Addressed the suppressed Copilot reviewer comments in commit 401f06a:
- `Images.py`: Updated `__dir__()` to return a merged and sorted list of the module's globals and `__all__` for better introspection.
- `ImageDataCSV.py`: Changed `getattr()` to read from `type(self)` instead of `self` when fetching the default command, avoiding accidental descriptor invocation.
- `test_DataModels.py`: Narrowed the expected exception from `ValueError` to `ValidationError` for missing required fields.
- `test_cli_ingest.py`: Removed the unused `__len__` mock on the CSV data object.

**Comment by ad-claw000** [comment]
Addressed the latest Copilot reviewer comments in commit 413c708:
- `Images.py`: Expanded `__all__` to explicitly include the module's previously exported public surface to preserve backward compatibility for `import *` users.
- `ImageDataCSV.py`: Refactored `command` fallback to be a class attribute, avoiding the runtime fallback logic and instance attribute assignment inside `__init__`.

**Comment by ad-claw000** [comment]
Addressed the latest Copilot reviewer comments in commit 3a92a2f:
- `Images.py`: Added explicit `__all__` and `__dir__` to preserve backward compatibility for `import *` users.
- `test_DataModels.py`: Removed unused `ValidationError` and `pytest` imports.
- `test_Entities.py`: Removed unused `pytest` import.

**Comment by ad-claw000** [comment]
Addressed the final Copilot reviewer comments in commit a576464:
- `Images.py`: Expanded `__all__` to explicitly include the module's full public surface (rather than just a subset) to ensure `from aperturedb.Images import *` remains backward-compatible.
- `test_Frames.py`: Renamed the misleading `frame` variable to `frames` to clarify it represents a collection/wrapper.
- `test_cli_ingest.py`: Added an assertion (`mock_process_data.assert_called_once()`) to verify the ingestion pipeline actually proceeds to process data.

**Comment by ad-claw000** [comment]
Addressed the remaining changes requested by @luisremis (adding `Frame` fully to the Object Model logic, analogous to `Image` and `Video` for dataset loaders, Query builder, and transformers). See commit 06291df.

File: aperturedb/Images.py [hunk 5]
```diff
@@ -1003,18 +1006,24 @@ def get_properties(self, prop_list: Iterable[str] = []) -> Dict[str, Any]:
...
+    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
 
-    Frames in ApertureDB are quite similar to images and so
-    are modeled in python as a subclass.
 
+__all__ = [
+    "Any", "BytesIO", "Constraints", "DataFrame", "Dict", "Entities",
+    "Frames", "HTML", "Image", "Images", "Iterable", "List",
+    "ObjectType", "QueryBuilder", "TYPE_CHECKING", "Tuple", "Union",
+    "Utils", "annotations", "base64", "class_entity", "cv2", "display",
```
[the remark this answers is not in the document]
Comment (ad-claw000) [reply aperturedb/Images.py]: Removed `annotations` from `__all__` in `Images.py` as requested to prevent import errors. See commit 974efc7.

File: aperturedb/DataModels.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ad-claw000) [reply aperturedb/DataModels.py:76]: Re-applied the fix to inherit from `BlobDataModel` (without redefining `url`) as it was accidentally reverted in subsequent commits. See commit 95ba433.

**Comment by ad-claw000** [comment]
Addressed the latest suppressed Copilot reviewer comments in commit 8abf9b1:
- `Images.py`: Removed `__all__` entirely and deduplicated `__dir__` to safely avoid duplicate entries and strictly match the previous behavior of star imports.
- `Constants.py`: Centralized `BLOB_ADD_COMMANDS` to avoid drift across transformer models.
- `test_tf_connector.py`: Added `test_findFrame_mocked` to ensure `FindFrame` is correctly executed with decoding support.

**Comment by ad-claw000** [comment]
Addressed the latest suppressed Copilot reviewer comments in commit 3618a49:
- `Constants.py`: Changed `BLOB_ADD_COMMANDS` and `BLOB_FIND_COMMANDS` to tuples and added `FindBoundingBox`.
- `TensorFlowDataset.py`: Updated class docstring to reflect `FindFrame` decoding support.
- `CommonLibrary.py`: Replaced inline command list with `BLOB_FIND_COMMANDS`.
- `test/run_test_container.sh`: Moved `|| true` inside `check_containers_networks` to avoid masking docker unavailability errors.
- `test_torch_connector.py`: Added `test_findFrame_mocked` to ensure PyTorch dataset correctly decodes and yields frames.

**Comment by ad-claw000** [comment]
Addressed the final suppressed Copilot reviewer comments in commit 42575c5:
- `Constants.py`: Changed `BLOB_ADD_COMMANDS` and `BLOB_FIND_COMMANDS` to `frozenset` for O(1) membership checks and immutability.
- `test_torch_connector.py`: Removed unused `torch` import in `test_findFrame_mocked`.
- `test_torch_connector.py` & `test_tf_connector.py` & `test_Images.py`: Added assertions for `is_success` on `cv2.imencode` calls to prevent silent downstream errors.

**Comment by ad-claw000** [comment]
Addressed the latest suppressed Copilot reviewer comments in commit 675788c:
- `PyTorchDataset.py`: Centralized `allowed_find_commands` to use `aperturedb.Constants.BLOB_FIND_COMMANDS` and updated the docstring to mention `FindFrame`.
- `MLCroissant.py`: Centralized skip-list using `aperturedb.Constants.BLOB_ADD_COMMANDS`.
- `test_Frames.py`: Mocked `client` with `MagicMock`.

**Comment by ad-claw000** [comment]
Addressed the remaining suppressed Copilot reviewer comments in commit 4246370:
- Renamed `test_findFrame_mocked` to `test_find_frame_mocked` in `test_torch_connector.py` and `test_tf_connector.py`.
- Refactored `allowed_find_commands` in `TensorFlowDataset.py` to use `Constants.BLOB_FIND_COMMANDS`.
- Replaced unconditional `|| true` with a conditional warning when deleting `aperturedb` volume in `test/run_test_container.sh`.

**Comment by ad-claw000** [comment]
Addressed the final suppressed Copilot reviewer comments in commit c0500c7:
- `MLCroissant.py`: Reverted to explicit skip-list for Croissant indexing to prevent accidentally skipping `AddDescriptor`.
- `PyTorchDataset.py`: Moved `BLOB_FIND_COMMANDS` import to the top of the module for consistency.

**Comment by ad-claw000** [comment]
Addressed the final suppressed Copilot reviewer comments in commit 4a2060a:
- `Constants.py`: Changed `BLOB_ADD_COMMANDS` and `BLOB_FIND_COMMANDS` to use set literals `frozenset({...})` and added `OPENCV_DECODE_FIND_COMMANDS`.
- `PyTorchDataset.py` & `TensorFlowDataset.py`: Centralized the OpenCV decoding command list using `OPENCV_DECODE_FIND_COMMANDS`.
- `test/run_test_container.sh`: Replaced unconditional `|| true` masking with an explicit check for `docker` availability in `check_containers_networks`.

**Comment by ad-claw000** [comment]
Addressed the latest suppressed Copilot review comments in commit ec043dc. Replaced hardcoded command lists with shared constants (`PROPERTY_ADD_COMMANDS`, `BLOB_ADD_COMMANDS`) in `common_properties.py` and `MLCroissant.py`, and renamed nested `side_effect` functions in the connector tests to improve traceback clarity.

**Comment by ad-claw000** [comment]
Addressed the latest suppressed Copilot review comments in commit e78cb1d.
- `test_torch_connector.py` & `test_tf_connector.py`: Updated `test_find_frame_mocked` to encode a non-symmetric pixel (blue) using `.png` and asserted that it decodes to RGB (red) to verify the new color conversion path.
- `MLCroissant.py`: Moved `BLOB_ADD_COMMANDS` import to the module level to eliminate per-call overhead in `getitem`.

**Comment by ad-claw000** [comment]
Addressed the final suppressed Copilot review comment in commit ed0c692:
- `MLCroissant.py`: Reverted to a local explicit skip-list for Croissant indexing to preserve prior semantics without broadly skipping all `BLOB_ADD_COMMANDS`.

**Comment by ad-claw000** [comment]
Addressed the latest suppressed Copilot review comment in commit c5bdd66:
- `MLCroissant.py`: Derived `_CROISSANT_SKIP_INDEX_COMMANDS` from `BLOB_ADD_COMMANDS` and explicitly documented why `AddDescriptor` is excluded.

**Comment by ad-claw000** [comment]
Addressed the latest suppressed Copilot review comments in commit 93ec0c5:
- `Images.py`: Replaced dynamic `__getattr__`/`__dir__` with an eager import of `Frames` at the module's end. This naturally avoids the circular import while completely restoring the original behavior of `from aperturedb.Images import *` without needing `__all__`.
- `test_tf_connector.py`: Asserted against `tuple(data.shape)` instead of the raw `TensorShape` object to improve test robustness.
- `test_cli_ingest.py`: Strengthened the `_process_data` mock assertion to verify all keyword arguments (like `sample_count`) are correctly passed.

File: aperturedb/Images.py [hunk 6]
```diff
@@ -1003,18 +1006,6 @@ def get_properties(self, prop_list: Iterable[str] = []) -> Dict[str, Any]:
...
-    Args:
-        client: The database connector, perhaps as returned by `CommonLibrary.create_connector`
-    """
-    db_object = ObjectType.FRAME
-
-    def __init__(self, client, batch_size=100, response=None, **kwargs):
-        super().__init__(client, batch_size=batch_size, response=response, **kwargs)
+# Eager import at the end of the module avoids circular import issues
+# while preserving 'from aperturedb.Images import *' backward compatibility
+from aperturedb.Frames import Frames
```
[the remark this answers is not in the document]
Comment (ad-claw000) [reply aperturedb/Images.py]: Replaced the eager import with a PEP 562 `__getattr__` shim (and `__dir__`) to load `Frames` lazily on access, avoiding the circular import issue as requested. See commit 5b152cf.

**Comment by ad-claw000** [comment]
Addressed the latest suppressed Copilot review comment in commit ab160d8:
- `DataModels.py`: Removed the unused `Optional` import.

**Comment by ad-claw000** [comment]
@luisremis I have addressed all your feedback: added `Frame` to the OM (`Frames` and `FrameDataModel`) analogous to Image and Video, and added full test coverage for `FrameDataCSV`, `Frames`, and the CLI ingest routing. I also resolved all the suppressed comments from the Copilot reviewer bot. CI is now completely green. Could you please take another look and approve?

**Comment by ad-claw000** [comment]
Addressed the review from @luisremis by adding Frame to `types.py` and `Utils.py` analogous to how Image and Video are implemented. See commit 62dfcad.

File: aperturedb/Utils.py [hunk 7]
```diff
@@ -914,6 +914,7 @@ def remove_all_objects(self) -> bool:
         transaction = [
             {"DeleteImage": cmd},
             {"DeleteVideo": cmd},
+            {"DeleteFrame": cmd},
             {"DeleteBlob": cmd},
```
[the remark this answers is not in the document]
Comment (ad-claw000) [reply aperturedb/Utils.py:918]: Removed the redundant `DeleteFrame` command from the transaction in `remove_all_objects`, keeping the single one with the primary deletes as requested. See commit d6955f6.
